### DISCUSSION TITLE
buffer close but quit on last buffer

### DISCUSSION THREAD

**Discussion by ghost** [helix-editor/helix#7719] [General] [answered]
I mapped 
`q` = `":buffer-close"`
`Q` = `":buffer-close!"`
but when i am pressing q while it is on last buffer, helix is creating a new scratch buffer removing the original one

**Comment by woojiq** [comment]
I think this is a useless improvement as there are other commands like `qa` for your purposes. Imagine that you are writing something in a editor with gui (for example, vscode). If you want to exit completely, you press the top `X` button and do not close all open files one by one. Same thing: `quit-all` is your GUI `X` button and `buffer-close` is the `X` next to the file name. You do not close the program, if you have closed all the files, the `welcome screen` just opens.

Also, if it was the last buffer and you didn't notice, you can simply open another file and the lsp will continue to work. However with your suggestion, you will have to wait tens of seconds again to wait the lsp.

**Reply by ghost** [reply]
Yeah I know, there is `quit-all` command but typing out commands is slow and boring
I use `z` for `write` and `q` for `quit` if I used the mannual way I would have used `esc` `:` `w` `q` `ret` and `a` sometimes to quit all

**Comment by gabydd** [comment]
The above explains why this is the intended behaviour well, I will convert this to discussion so you can discuss alternatives but I don't think we want to change the behaviour of ``buffer-close``, those commands are for managing open buffers changing it shouldn't quit the editor that's better left for other commands

**Reply by ghost** [reply, marked as answer]
I am not saying to change the behaviour of `buffer-close`, just add another command like `buffer-close-and-quit` and also same for force quit `buffer-close-and-quit!`. It will work very similar to less command if you map `buffer-close-and-quit` to `q`

**Reply by Cloudef** [reply]
I'd love to have this too. Just recently moved to hx from vim/neovim and even on there having this feature cleanly is annoying (but doable).

**Comment by EnzoMoser** [comment]
Here is the nvim plugin implementation: 

https://github.com/psjay/buffer-closer.nvim

https://private-user-images.githubusercontent.com/856285/349017790-067f2503-2e00-4dcc-bc97-d1b74d544b1a.mp4?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzMxNjk4NzgsIm5iZiI6MTc3MzE2OTU3OCwicGF0aCI6Ii84NTYyODUvMzQ5MDE3NzkwLTA2N2YyNTAzLTJlMDAtNGRjYy1iYzk3LWQxYjc0ZDU0NGIxYS5tcDQ_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMxMFQxOTA2MThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZTU1MzllMDA0MTc5NGFkMGNkZDFhYmU1ZmU3M2M0MTVkYTk5NzRiOTBlYzU4NGNmZDg5OWI3M2Q0ZGE5MzQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.dfQxC5XZ4OtDJaluFPtlbH8H2ACzGFZ69bQXFNKIiLw

Personally, I think having a command that exits on the last buffer makes sense. It falls in line with exiting when closing the last window.

**Comment by Cloudef** [comment]
Fyi I made PR for this feature here <https://github.com/helix-editor/helix/pull/15311>
